src/back/spv/writer.rs

```diff
@@ -1320,6 +1355,234 @@ impl Writer {
+        block.body.push(Instruction::array_length(
+            length_type_id,
+            length_id,
+            structure_id,
+            member_idx,
+        ));
+
+        Ok(length_id)
+    }
+
+    /// Emit code to compute the maximum valid index of a run-time array.
+    ///
+    /// Given `array`, an expression referring to the final member of a struct,
+    /// where the member in question is a runtime-sized array, return the id of
+    /// a SPIR-V insn representing the maximum valid index into that array (that
+    /// is, one less than its length).
+    fn write_runtime_array_limit(
+        &mut self,
+        array: Handle<crate::Expression>,
+        ir_module: &crate::Module,
+        ir_function: &crate::Function,
+        function: &Function,
+        block: &mut Block,
+    ) -> Result<Word, Error> {
+        let length_id =
+            self.write_runtime_array_length(array, ir_module, ir_function, function, block)?;
+
+        let limit_type = LocalType::Value {
+            vector_size: None,
+            kind: crate::ScalarKind::Uint,
+            width: 4,
+            pointer_class: None,
+        };
+        let limit_type_id = self.get_type_id(limit_type.into())?;
+        let const_one_id = self.get_index_constant(1)?;
+
+        let limit_id = self.id_gen.next();
+        block.body.push(Instruction::binary(
+            spirv::Op::ISub,
+            limit_type_id,
+            limit_id,
+            length_id,
+            const_one_id,
+        ));
+
+        Ok(limit_id)
+    }
+
+    /// Emit code to restrict an index to be in range for a vector, matrix, or array.
+    ///
+    /// Add instructions to `block` to restrict `index_id` to the proper range
+    /// for indexes into a value of type `base_ty`. Return the id of the
+    /// restricted index value.
+    ///
+    /// By 'restricted', we mean only that in-bounds indices are unchanged, and
+    /// out-of-bounds indices are changed to some arbitrary in-bounds index.
+    /// This is not necessarily clamping; for example, negative indices might be
+    /// changed to refer to the last element of the array, not the first, as
+    /// clamping would do.
+    ///
+    /// If `base_ty`'s length is not known at compile time, just return `index`
+    /// unclamped. TODO: find the dynamic length.
+    #[allow(clippy::too_many_arguments)]
+    fn write_restricted_index(
+        &mut self,
+        array: Handle<crate::Expression>,
+        index: Handle<crate::Expression>,
+        ir_module: &crate::Module,
+        ir_function: &crate::Function,
+        fun_info: &FunctionInfo,
+        function: &Function,
+        block: &mut Block,
+    ) -> Result<Word, Error> {
+        let array_ty = fun_info[array].ty.inner_with(&ir_module.types);
+        let index_id = self.cached[index];
+        let index_type_id = self.get_expression_type_id(&fun_info[index].ty)?;
+
+        let limit_id = if let Some(known_length) = array_ty.to_known_length(ir_module) {
```

Comment:
nit: doing `if let ... = ... { ... } else {..}` is a weird substitute for `match`

Reply:
The nice thing about `if let else` is that it saves you a level of indentation, compared to match. But I can make this a match if you like.